README.md
# appa

Quickly create simple JSON API services.

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![standard][standard-image]][standard-url]
[![conduct][conduct]][conduct-url]

[npm-image]: https://img.shields.io/npm/v/appa.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/appa
[travis-image]: https://img.shields.io/travis/sethvincent/appa.svg?style=flat-square
[travis-url]: https://travis-ci.org/sethvincent/appa
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[standard-url]: http://npm.im/standard
[conduct]: https://img.shields.io/badge/code%20of%20conduct-contributor%20covenant-green.svg?style=flat-square
[conduct-url]: CONDUCT.md

![appa](https://raw.githubusercontent.com/sethvincent/appa/master/appa.jpg)

## Install

Make sure you've got [node installed](http://nodejs.org), then make `appa` a project dependency:

```sh
npm install --save appa
```

## Usage

```js
var http = require('http')
var app = require('appa')()
var send = require('appa/send')
var log = app.log

app.on('/', function (req, res, context) {
  send({ message: 'oh hey friends' }).pipe(res)
})

http.createServer(app).listen(3000, function () {
  log.info('server started at http://127.0.0.1:3000')
})
```

## Error handling

Any uncaught errors that occur in a request handler will be caught and a `500 Internal server error` response will be sent.

Send error responses using the `appa/error` module:

```js
var error = require('appa/error')

module.exports = function (req, res, ctx) {
  return error(404, 'Not found').pipe(res)
}
```

Sending an error response does not automatically log the error, so to add that you can do something like:

```js
var error = require('appa/error')
var log = require('appa/log')()

module.exports = function (req, res, ctx) {
  log.error(req.method, '500', errorStack)
  return error(500, 'Internal server error').pipe(res)
}
```

## Logging

appa uses [pino](https://npmjs.com/pino) for logging. Pass options to pino with `options.log`: `appa({ log: pinoOptions })`.

See [example pino usage](https://github.com/pinojs/pino#usage) and [all pino options](https://github.com/pinojs/pino/blob/master/docs/API.md#pinooptions-stream).

Or disable logging completely by setting `options.log` to `false`: `appa({ log: false })`.

## Documentation
- [API](docs/api.md)
- [Tests](tests/)

### Examples
- [Basic example](examples/basic.js)
- Example [server](examples/server.js) and [client](examples/client.js)

## Contributing

Contributions are welcome! Please read the [contributing guidelines](CONTRIBUTING.md) first.

## Conduct

It is important that this project contributes to a friendly, safe, and welcoming environment for all. Read this project's [code of conduct](CONDUCT.md)

## Changelog

Read about the changes to this project in [CHANGELOG.md](CHANGELOG.md). The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Contact

- **issues** – Please open issues in the [issues queue](https://github.com/sethvincent/appa/issues)
- **twitter** – Have a question? [@sethdvincent](https://twitter.com/sethdvincent)
- **email** – Need in-depth support via paid contract? Send an email to sethvincent@gmail.com

## License

[ISC](LICENSE.md)


CHANGELOG.md
# appa change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

> Note: unreleased changes are listed here

## 6.1.3 - 2017-07-02

## Fixed

- make it possible to skip body parsing with `{ parse: false }` option

## 6.1.2 - 2017-02-25

## Fixed

- fix double response bug

## 6.1.1 - 2017-02-25

## Fixed

- uncaught errors inside request handlers are now caught and a `500 Internal server error` response is sent
- Added docs to readme about error handling and logging
- update tests/README.md

## 6.1.0 - 2017-02-19

## Added
- disable logging with `var app = appa({ logging: false })`
- Improved [api docs](docs/api.md)
- refactored to use the raw-body module, which allows setting the limit of incoming request body sizes
- individual request handlers can now pass a `parseJSON: false` option to disable JSON parsing
- add [server](examples/server.js) and [client](examples/client.js) examples

## 6.0.0 - 2016-11-01

### Changed
- use `send(data).pipe(res)` rather than `send(res, data)`

## 5.0.1 - 2016-10-24

### Added
- Add a docs directory
- Add CONDUCT.md
- Add CHANGELOG.md

### Changed
- Move send, error, and log into separate files
- Revise README.md with links and updated info.
- License: MIT > ISC
- Create tests directory


index.js
var assert = require('assert')
var qs = require('qs')
var url = require('url')
var rawBody = require('raw-body')
var createRouter = require('wayfarer')
var isType = require('type-is')
var httplogger = require('pino-http')

var createLogger = require('./log')
var send = require('./send')
var error = require('./error')

/**
* Create the application. Returns the `app` function that can be passed into `http.createServer`.
* @name createApp
* @param {Object} config
* @param {Object|Boolean} config.log – Set to `false` to disable logging. Set to an object with options that are passed to [pino](https://npmjs.com/pino) and [pino-http](https://npmjs.com/pino-http)
* @param {Function} config.notFound – Function that handlers 404 routes. Provides `req`, `res`, and `ctx` arguments just like other appa route handlers. Default: a function that sends a `404` statusCode with the message `Not found`.
* @returns {Function} `app`
* @example
* var appa = require('appa')
*
* var app = appa({
*   log: { level: 'info' }, // or set to `false` to disable all logging
*   notFound: function (req, res, ctx) {
*     return app.error(404, 'Not Found').pipe(res)
*   }
* })
*/
module.exports = function createApp (config) {
  config = config || {}

  config.log = config.log === false
    ? { level: 'silent' }
    : config.log || { level: 'info' }

  var log = createLogger(config.log)
  var httplog = httplogger(config.log, config.log.stream)
  var router = app.router = createRouter('/404')

  // provide a 404 fallback
  on('/404', (config.notFound || notFound))
  function notFound (req, res) { error('Not found').pipe(res) }

  // ignore favicon.ico requests
  on('/favicon.ico', function (req, res) { send(200).pipe(res) })

  /**
  * The request, response handler that is passed to `http.createServer`, and the object that
  * provides methods for your app.
  * @name app
  * @param {Object} req – the http request object
  * @param {Object} res – the http response object
  * @example
  * var http = require('http')
  * var appa = require('appa')
  *
  * var app = appa()
  * var server = http.createServer(app)
  */
  function app (req, res) {
    httplog(req, res)
    var ctx = url.parse(req.url)
    ctx.query = qs.parse(ctx.query)
    return router(ctx.pathname, req, res, ctx)
  }

  /**
  * Route handler
  * @name app.on
  * @param {String} pathname – the route for this handler
  * @param {Object} options – options for the request handler
  * @param {Boolean} options.parse – optionally disable parsing of the body. Default: `true`
  * @param {Function} callback – the route handler
  * @example
  * app.on('/', function (req, res, ctx) {
  *   // handle the route
  * })
  */
  function on (pathname, options, callback) {
    if (typeof options === 'function') {
      callback = options
      options = {}
    }

    assert.equal(typeof pathname, 'string', 'appa: pathname is required and must be a string')
    assert.equal(typeof options, 'object', 'appa: options must be an object')
    assert.equal(typeof callback, 'function', 'appa: callback function is required')

    options.parse = options.parse === false ? options.parse : true

    return router.on(pathname, function (params, req, res, ctx) {
      ctx.params = params
      log.info(ctx)

      function parse (req, options) {
        body(req, options, function (err, result) {
          if (err) return error(err.status, err.type).pipe(res)

          if (isType(req, ['json'])) {
            ctx.body = parseJSON(res, result)
          } else {
            ctx.body = result
          }

          return respond(req, res, ctx)
        })
      }

      function respond (req, res, ctx) {
        try {
          return callback(req, res, ctx)
        } catch (e) {
          return error(500, 'Internal server error', e).pipe(res)
        }
      }

      if (req.method === 'POST' || req.method === 'PUT' || req.method === 'PATCH') {
        if (options.parse === false) return respond(req, res, ctx)
        return parse(req, options)
      }

      return respond(req, res, ctx)
    })
  }

  function parseJSON (res, result) {
    try {
      return JSON.parse(result)
    } catch (err) {
      return error(400, 'Invalid JSON').pipe(res)
    }
  }

  function body (req, options, callback) {
    if (!options.limit) options.limit = '1mb'
    if (!options.encoding) options.encoding = 'utf8'
    rawBody(req, options, callback)
  }

  /**
  * Send a JSON object as a response
  * @name app.send
  * @param {Number} statusCode – the status code of the response, default is 200
  * @param {Object} data – the data that will be stringified into JSON
  * @example
  * var send = require('appa/send')
  *
  * app.on('/', function (req, res, ctx) {
  *   send({ message: 'hi' }).pipe(res)
  * })
  */
  app.send = send

  /**
  * Send a JSON error response
  * @name app.error
  * @param {Number} statusCode – the status code of the response, default is 404
  * @param {String} message – the message that will be stringified into JSON
  * @param {Object} data – additional data about the error to send in the response
  * @example
  * var error = require('appa/error')
  *
  * app.on('/', function (req, res, ctx) {
  *   error(404, 'Resource not found').pipe(res)
  * })
  */
  app.error = error

  /**
  * Create logs using the pino module: https://npmjs.com/pino
  * @name app.log
  */
  app.log = log

  /**
  * Parse or stringify a JSON stream using the JSONStream module: https://npmjs.com/JSONStream
  * @name app.json
  */
  app.json = require('JSONStream')

  /**
  * Compose a stream using the pump module: https://npmjs.com/pump
  * @name app.pipe
  */
  app.pipe = require('pump')

  app.on = on
  return app
}
